corehq/apps/data_interfaces/static/data_interfaces/js/case_dedupe_main.js

```diff
@@ -14,19 +14,156 @@ hqDefine("data_interfaces/js/case_dedupe_main", [
     casePropertyInput,
     CaseRuleCriteria
 ) {
+
+    /*
+    PropertyManager exists to make sure that available properties are updated and unique.
+    For example, if a case type contains A, B, and C properties, this class ensures
+    that if C is selected, further options will only show A and B.
+    */
+    const PropertyManager = function (caseType, selectedProperties, initialPropertyMap) {
+        const self = {};
+        self.caseType = caseType;
+        self.selectedProperties = selectedProperties;
+        self.availablePropertyMap = ko.observable(clonePropertyMap(initialPropertyMap));
+
+        self.addProperty = function (property) {
+            self.selectedProperties.push(property);
+
+            if (property.name()) {
+                let currentProperties = self.availablePropertyMap()[self.caseType()];
+                currentProperties = reserveProperties([property.name()], currentProperties);
+                rebuildPropertyMap(currentProperties);
+            }
+        };
+
+        self.removeProperty = function (property) {
+            self.selectedProperties.remove(property);
+
+            let currentProperties = self.availablePropertyMap()[self.caseType()];
+            currentProperties = restoreProperty(property.name(), currentProperties);
+
+            rebuildPropertyMap(currentProperties);
+        };
+
+        self.updatePropertyValue = function (oldValue, newValue) {
+            let currentProperties = self.availablePropertyMap()[self.caseType()];
+            currentProperties = reserveProperties([oldValue], currentProperties);
+            currentProperties = restoreProperty(newValue, currentProperties);
```

Comment:
If we're understanding right, should the new value be "reserved" (not available for selection) and the old value be "restored" (available for selection again)?

Reply:
Correct. Because `PropertyManager` also exposes its own add and remove functionality, it was hard to come up with precise names for this behavior.

Reply:
I think the confusion here was, should we not reserve the `newValue` and restore the `oldValue`?
Something like - 
```
reserveProperties([newValue], currentProperties);
restoreProperty(oldValue, currentProperties);
```

Reply:
Ah, I was feeding in the properties in reverse order, so great catch! Addressed in c8c829ee236e7d6998449c8c4a06304bebe50e86